File: macos/Sources/CredentialProxy/Services/DomainMatcher.swift
import Foundation

enum DomainMatcher {
    /// Check if a domain matches an allowed pattern.
    /// Supports exact match and wildcard subdomain (`*.example.com` matches `sub.example.com` but not `example.com`).
    static func matchesDomain(_ domain: String, pattern: String) -> Bool {
        let domain = domain.lowercased()
        let pattern = pattern.lowercased()

        if domain == pattern {
            return true
        }

        if pattern.hasPrefix("*.") {
            let suffix = String(pattern.dropFirst(1)) // ".example.com"
            return domain.hasSuffix(suffix) && domain.count > suffix.count
        }

        return false
    }

    /// Check if a domain is allowed by any pattern in the list.
    static func isDomainAllowed(_ domain: String, allowedDomains: [String]) -> Bool {
        allowedDomains.contains { matchesDomain(domain, pattern: $0) }
    }

    /// Extract the host from a URL string.
    static func extractDomain(from urlString: String) -> String? {
        URL(string: urlString)?.host
    }
}

File: src/utils/domain-matcher.ts
/**
 * Check if a domain matches an allowed pattern
 * Supports:
 * - Exact match: "api.linear.app"
 * - Wildcard subdomain: "*.linear.app" (matches any subdomain, but not the root)
 */
export function matchesDomain(domain: string, pattern: string): boolean {
  // Normalize to lowercase
  domain = domain.toLowerCase();
  pattern = pattern.toLowerCase();

  // Exact match
  if (domain === pattern) {
    return true;
  }

  // Wildcard match
  if (pattern.startsWith('*.')) {
    const suffix = pattern.slice(1); // ".linear.app"
    return domain.endsWith(suffix) && domain.length > suffix.length;
  }

  return false;
}

/**
 * Check if a domain is allowed by any pattern in the list
 */
export function isDomainAllowed(domain: string, allowedDomains: string[]): boolean {
  return allowedDomains.some(pattern => matchesDomain(domain, pattern));
}

/**
 * Extract domain from a URL
 */
export function extractDomain(url: string): string {
  try {
    const parsed = new URL(url);
    return parsed.hostname;
  } catch {
    throw new Error(`Invalid URL: ${url}`);
  }
}
